__tests__/atomWithQueryParam_spec.tsx

```diff
@@ -0,0 +1,121 @@
+import { act, renderHook } from '@testing-library/react';
+import { Provider, useAtom } from 'jotai';
+import { RESET } from 'jotai/utils';
+
+import { atomWithQueryParam } from '../src/atomWithQueryParam';
+
+let pushStateSpy: jest.SpyInstance;
+let replaceStateSpy: jest.SpyInstance;
+
+beforeEach(() => {
+  pushStateSpy = jest.spyOn(window.history, 'pushState');
+  replaceStateSpy = jest.spyOn(window.history, 'replaceState');
+});
+
+afterEach(() => {
+  pushStateSpy.mockRestore();
+  replaceStateSpy.mockRestore();
+});
+
+describe('atomWithQueryParam', () => {
+  it('should return a default value for the atom if the query parameter is not present', () => {
+    const queryParamAtom = atomWithQueryParam('test', 'default');
+    const { result } = renderHook(() => useAtom(queryParamAtom), {
+      // the provider scopes the atoms to a store so their values dont persist between tests
+      wrapper: Provider,
+    });
+    expect(result.current[0]).toEqual('default');
+  });
+
+  it('should sync an atom to a query parameter', () => {
+    const queryParamAtom = atomWithQueryParam('test', {
+      value: 'default',
+    });
+    const { result } = renderHook(() => useAtom(queryParamAtom), {
+      // the provider scopes the atoms to a store so their values dont persist between tests
+      wrapper: Provider,
+    });
+
+    act(() => {
+      result.current[1]({ value: 'test value' });
+    });
+
+    expect(result.current[0]).toEqual({ value: 'test value' });
```

Comment:
I really like these tests, but feel that they could be a bit more verbose and/or easy to read. I feel tests can help act as a sort of documentation in a way.

What follows is just a quick rewrite as an example, I haven't tested it out locally.
```suggestion
    const { result } = renderHook(() => useAtom(queryParamAtom), {
      // the provider scopes the atoms to a store so their values dont persist between tests
      wrapper: Provider,
    });
    const [testValue, setTestValue] = result.current;

    act(() => {
      setTestValue({ value: 'test value' });
    });

    expect(testValue).toEqual({ value: 'test value' });
```

Reply:
`const [testValue, setTestValue] = result.current;`

i would expect this change to break the test actually because RTL updates result.current with the new testValue but that wouldn't rerun the destructuring after the act call. 

That's the reason for the unfortunate index access syntax (which Jotai even has [in it's docs](https://jotai.org/docs/guides/testing#custom-hooks))